### PR TITLE
feat: ondevice addon controls v1 [6.0]

### LINKED ISSUE #65: [6.0] support for declarative config like web storybook
**Is your feature request related to a problem? Please describe.**

Looks like web version supports declarative config but react native one still is in old config. I was wondering whether there is any plan to support declarative config here too? It will be lot easy to share configs for people who use monorepos with web and mobile apps.

**Describe the solution you'd like**

```
module.exports = {
  stories: [`${packagesPath}/**/*.stories.js`],
  addons: [
    '@storybook/addon-actions/register',
  ],
};
```
this web version's config will work with native too.

**Describe alternatives you've considered**
https://storybook.js.org/docs/guides/guide-react-native/

**Are you able to assist bring the feature to reality?**
no

### LINKED ISSUE #164: [6.0] Auto import stories
Implement auto importing for stories for 6.0.

Currently there is a basic setup with a pre-script based on a similar idea as the react native storybook loader.

Some things were attempted with pre-val and using a custom transformer in metro however we encountered issues with caching and things of that nature.

We should investigate if there is another option or continue with the pre-script approach and make it a binary for react-native storybook.

### LINKED ISSUE #177: [6.0] ondevice-controls
Now that args and CSF are supported its time to add in controls support.

Controls will 
- allow for changing props in real time similar to knobs
- detect prop types automatically
- utilise args to provide a simpler api than knobs

### PR DESCRIPTION
**Pull request by dannyhw**
Issue: #177 

## What I did

I've started converting knobs into controls.

Things missing
- ~~fix hooks context or make a custom one, hooks context from storybook/api hooks use the window object~~
- ~~currently the state only updates when until the preview is refocused~~ hacky fix added
- ~~"reset" doesn't reset the control state correctly~~
- ~~add back in all the other types~~ will be continued in future PR's
- ~~big refactor needed~~ partially done, will also be part of future work.

After this PR is merged this issue #181 should be picked up to improve the re-usability of these components.

## How to test

testing steps to come.

### CODE REVIEW

**Comment by dannyhw** [comment]
@lauriharpf hey I was looking to see whats left to do here in order to merge and I guess it's not quite ready.  However I think that there is a way to split some of this work.

if you see my commit https://github.com/storybookjs/react-native/pull/182/commits/2bd4f831fcaf5b2a6d855738d9ca552973454bb8

I've started re-doing the old knob components to work with controls, if you want to give it a try then pick one of the control types to add back, boolean is probably an easy one to start with. 

Also currently the examples are a bit useless so at some point it would be good to improve those, though I think the functionality is the first priority.

**Comment by lauriharpf** [comment]
> @lauriharpf hey I was looking to see whats left to do here in order to merge and I guess it's not quite ready. However I think that there is a way to split some of this work.
> 
> if you see my commit [2bd4f83](https://github.com/storybookjs/react-native/commit/2bd4f831fcaf5b2a6d855738d9ca552973454bb8)
> 
> I've started re-doing the old knob components to work with controls, if you want to give it a try then pick one of the control types to add back, boolean is probably an easy one to start with.
> 
> Also currently the examples are a bit useless so at some point it would be good to improve those, though I think the functionality is the first priority.

Thanks :+1:. Gave this a shot at https://github.com/storybookjs/react-native/pull/195 , please take a peek and see if it is in the right direction. This area isn't too familiar for me, so included some questions.

I'll need to tackle some other stuff for a week or so, which might make me a bit hard to reach. Should be back in action on the 12th of July or so. It might be possible for me to squeeze out another quick PR today or tomorrow, e.g. for the `SelectType`. Will of course incorporate any comments you might have! If I don't get it done by tomorrow then please just carry on & let's re-sync in a week or so.

**Comment by dannyhw** [comment]
@lauriharpf thanks again for your help, I really appreciate it. I will test your PR soon, if you need anything reach me here or via discord 🙂 .

**Comment by lauriharpf** [comment]
> It might be possible for me to squeeze out another quick PR today or tomorrow, e.g. for the `SelectType`. Will of course incorporate any comments you might have! If I don't get it done by tomorrow then please just carry on & let's re-sync in a week or so.

Shared where I ended up in https://github.com/storybookjs/react-native/pull/197 ; not much progress unfortunately. Feel free to pick it up or throw it away & just implement it 🙂 .

**Comment by lauriharpf** [comment]
@dannyhw , I'm back online again. Off the top of your head, anything I could pick up to help out?

**Comment by dannyhw** [comment]
> @dannyhw , I'm back online again. Off the top of your head, anything I could pick up to help out?

Hey @lauriharpf! yeah actually I'm thinking to merge this PR and break the rest of the work into new issues/tasks.

If you feel up to it then a review of the current changes would be really great. However I'm aware there is a lot of changes 😅 .


Once this is merged I'll create issues for the remaining controls work which you're welcome to pick up.

The remaining controls to re-add are
- ArrayType
- DateType
- ButtonType
- RadioType

After those are done I'm thinking to release the first real v6 alpha, I expect there might be a lot of issues with this alpha but it would help with testing the changes.

Then future iterations on controls will include the following:
- grouping like knobs had (maybe)
- create issues for any missing functionality when compared with web
- refactoring 
- select control mapping functionality (i left it out for now because it seems like a bigger rewrite would be needed)
- resolving the re-render issue properly
- move code into a lib package for re-usability and maybe include stories for the controls in the example storybook. 
- unit tests / functional tests - a lot of work needed so more of a nice to have

File: app/react-native/package.json
Comment (lauriharpf) [inline app/react-native/package.json:34]: Wondering if e.g. https://www.npmjs.com/package/shx would be useful here, just to keep the release process as platform-independent as possible 🤔 . I'm on a platform where `rm -rf` works fine, but the wider developer audience stuff works for the better.

Whatever we do would be reasonable to be in sync with e.g. `addons/ondevice-controls/package.json` and `addons/ondevice-notes/package.json` (either make all of them cross-platform or keep all as-is). If we change this, then it could be done in a separate issue.

File: addons/ondevice-controls/package.json
Comment (lauriharpf) [inline addons/ondevice-controls/package.json:29]: Trivial, but would copying just the `*.png`'s be enough? The `xcf` doesn't seem to be require'd, just a master file for the `color-circle.png`?

File: addons/ondevice-backgrounds/src/BackgroundPanel.tsx
Comment (lauriharpf) [inline addons/ondevice-backgrounds/src/BackgroundPanel.tsx:1]: Great that in addition to the features, you've also done a lot of small code quality improvements in this PR (e.g. removing these unnecessary disables and using `StyleSheet.create` instead of inlining styles) :+1: 🥇 .

File: examples/native/components/ControlExamples/Boolean/Boolean.stories.tsx
Comment (lauriharpf) [inline examples/native/components/ControlExamples/Boolean/Boolean.stories.tsx:1]: Looks like this & the other controls stories aren't picked up in the examples. Should we change `examples/native/storybook.requires.js` to contain

```
const getStories = () => {
  return [
    require('./components/PromiseTest/Button.stories.tsx'),
    require('./other_components/AnotherButton/AnotherButton.stories.tsx'),
    require('./components/ControlExamples/Boolean/Boolean.stories.tsx'),
    require('./components/ControlExamples/Color/Color.stories.tsx'),
    require('./components/ControlExamples/Number/Number.stories.tsx'),
    require('./components/ControlExamples/Object/Object.stories.tsx'),
    require('./components/ControlExamples/Select/Select.stories.tsx'),
    require('./components/ControlExamples/Text/Text.stories.tsx'),
  ];
};
```

or is there a better way to handle it?

File: examples/native/components/ControlExamples/Boolean/Boolean.tsx
Comment (lauriharpf) [inline examples/native/components/ControlExamples/Boolean/Boolean.tsx:23]: This behaves a bit weird for me on the Android emulator:

1. When I initially open it, everything looks fine:

Screen Shot 2021-07-11 at 19 30 38

2. If I then go to "Addons -> Controls" and toggle it off, it still renders fine:

Screen Shot 2021-07-11 at 19 32 53
 
3. If I again go to "Addons -> Controls" and toggle it back on, the rendering becomes wonky:

Screen Shot 2021-07-11 at 19 31 11

This doesn't happen on iOS. Explicitly setting the color for both off & on seems to fix this:

```
const styles = StyleSheet.create({
...
  onText: { color: 'black' },
  offText: { color: 'white' },
});
...
    <Text style={on ? styles.onText : styles.offText}>{on ? 'on' : 'off'}</Text>
```

File: examples/native/components/ControlExamples/Color/Color.tsx
Comment (lauriharpf) [inline examples/native/components/ControlExamples/Color/Color.tsx:12]: Trivial, but we might want to not use `FunctionComponent` or `FC`, as the React core team seems to lean on the side of "don't use it": https://github.com/facebook/create-react-app/pull/8177

File: app/react-native/package.json
Comment (dannyhw) [reply app/react-native/package.json:34]: I think they should all work platform independent, this was more of a temporary solution, same as the resources copy. I'm happy to go with a package like this one to resolve it. 

Also maybe some work is needed on the bootstrap and Dev script to make sure everything works together.

File: examples/native/components/ControlExamples/Number/Number.stories.tsx
Comment (lauriharpf) [inline examples/native/components/ControlExamples/Number/Number.stories.tsx:14]: This behaves in a strange way for me on the Android Emulator, can you reproduce it? Seems to work fine on the iOS emulator, can't reproduce the issue there.

1. Open the story Navigator.
2. Select the "Basic" story.
3. Open "Addons -> Controls"
4. Click on the "first" field so that the keyboard comes up.
5. Click back to "Navigator"

## What happens
The story list in the navigator gets truncated a lot (seems to only occupy ~50% or so of the screen) after Step 5.

## Expected result
The story list would be rendered in the same way in Step 5 as it was in Step 1.

### Screenshot after Step 1
Screen Shot 2021-07-11 at 20 08 59

### Screenshot after Step 4
Screen Shot 2021-07-11 at 20 11 45

### Screenshot after Step 5
Screen Shot 2021-07-11 at 20 13 15

File: addons/ondevice-controls/package.json
Comment (dannyhw) [reply addons/ondevice-controls/package.json:29]: Yeah just the pngs. The color picker ideally needs to be extracted from this package or a pr made to the original source to make that package usable.

File: examples/native/components/ControlExamples/Boolean/Boolean.stories.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Boolean/Boolean.stories.tsx:1]: Well the requires file is auto generated, it should work if the regex is correct in the main.js.

File: examples/native/components/ControlExamples/Boolean/Boolean.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Boolean/Boolean.tsx:23]: Oh I guess there isn't a default color on Android then. I'll change this. Thanks!

File: examples/native/components/ControlExamples/Color/Color.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Color/Color.tsx:12]: I'm fine with either way. Happy to remove it 👍.

File: examples/native/components/ControlExamples/Boolean/Boolean.stories.tsx
Comment (lauriharpf) [reply examples/native/components/ControlExamples/Boolean/Boolean.stories.tsx:1]: Good point! One of these days I might want to start reading comments 😅 :

```
/*
  do not change this file, it is auto generated by storybook. 
*/
```

The regexes are

```
  stories: [
    './components/*/*.stories.?(ts|tsx|js|jsx)',
    './other_components/AnotherButton/AnotherButton.stories.tsx',
  ],
```

So we could change the upper one to `'./components/**/*.stories.?(ts|tsx|js|jsx)',` to account for the deeper directory structure in `ControlExamples` - or group the control stories under `Controls` and have a flatter directory structure, whichever we want.

File: examples/native/components/ControlExamples/Color/Color.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Color/Color.tsx:12]: I guess we should remove FC / FunctionComponent type annotations across the rest of the codebase just to be consistent. I'm fine with just including the prop types and leaving the function type implicit so like:

`const MyComponent = ({X}:Props) => <View/>`

File: examples/native/components/ControlExamples/Number/Number.stories.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Number/Number.stories.tsx:14]: That's weird, I think it's related to the keyboard aware component. I'm not able to check now but I'll have a look tomorrow.

File: examples/native/components/ControlExamples/Color/Color.tsx
Comment (lauriharpf) [reply examples/native/components/ControlExamples/Color/Color.tsx:12]: Sounds great! We could tackle this in a separate PR, e.g. by running https://github.com/gndelia/codemod-replace-react-fc-typescript on the codebase.

File: .eslintignore
Comment (lauriharpf) [inline .eslintignore]: If you can brief me a bit on the background of these excludes, then it would be awesome :+1: . `prettier.config.js` and `tsconfig.json` seem to lint OK as-is, even if we don't ignore them 🤔 .

`scripts/` excludes both `/scripts` and `app/react-native/scripts`. Both have files with `/* eslint-disable */`s, which looks a bit confusing (as they are now excluded from linting anyhow). Instead of excluding both `/scripts` directories completely, could we fix the eslint errors?

Made a small PR, https://github.com/storybookjs/react-native/pull/209 , that contains this & a simple fix for the regex thing we discussed in https://github.com/storybookjs/react-native/pull/182#discussion_r667504748

Comment (dannyhw) [reply .eslintignore]: I was battling with a lot of linting issues so this is probably just a result of that. Any improvements are welcomed so thanks for the fix 🙏 . Merged your pr already.

File: addons/ondevice-controls/src/GroupTabs.tsx
Comment (lauriharpf) [inline addons/ondevice-controls/src/GroupTabs.tsx:1]: I wonder where `GroupTabs` is used 🤔 . Searching the codebase for "GroupTabs" didn't seem to turn up anything. Tried deleting the file and didn't notice anything breaking. Do you know the purpose of the component?

File: addons/ondevice-controls/src/Panel.tsx
Comment (lauriharpf) [inline addons/ondevice-controls/src/Panel.tsx:5]: Could these commented-out portions be removed? 🤔

File: addons/ondevice-controls/src/PropForm.tsx
Comment (lauriharpf) [inline addons/ondevice-controls/src/PropForm.tsx:1]: Trivial, but `yarn lint` seems to pass even after removing this 🤔

**Review by lauriharpf — APPROVED** [review]
Phew! What an effort, well done! :+1: 🚀  🥇 

Tried to look at everything and also tested the controls on both platforms. IMO, we could merge this now and then tackle the remaining tasks in separate PRs. Let's try to split the work a bit today and look at how we could get the first alpha out.

Again, thank you 🙏  for this huge step towards v6! 💎

File: addons/ondevice-controls/src/GroupTabs.tsx
Comment (lauriharpf) [reply addons/ondevice-controls/src/GroupTabs.tsx:1]: Ah, I wonder if it's related to what you mentioned here: https://github.com/storybookjs/react-native/pull/182#issuecomment-877669981

> Then future iterations on controls will include the following:
> * grouping like knobs had (maybe)

Comment (dannyhw) [reply addons/ondevice-controls/src/GroupTabs.tsx:1]: Yeah you can see it here in the old knobs implementation.
https://github.com/storybookjs/react-native/blob/next-6.0/addons/ondevice-knobs/src/panel.js

File: app/react-native/package.json
Comment (dannyhw) [reply app/react-native/package.json:34]: Lets tackle this when we are closer to the alpha release.

File: addons/ondevice-controls/package.json
Comment (dannyhw) [reply addons/ondevice-controls/package.json:29]: To be tackled later

File: examples/native/components/ControlExamples/Color/Color.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Color/Color.tsx:12]: We will open an issue to run this codemod

File: examples/native/components/ControlExamples/Number/Number.stories.tsx
Comment (dannyhw) [reply examples/native/components/ControlExamples/Number/Number.stories.tsx:14]: Will create an issue to tackle this bug, seems to be for all text inputs on knobs/controls

File: addons/ondevice-controls/src/GroupTabs.tsx
Comment (dannyhw) [reply addons/ondevice-controls/src/GroupTabs.tsx:1]: we should see if the current controls implementation on the web has groups

File: addons/ondevice-controls/src/Panel.tsx
Comment (dannyhw) [reply addons/ondevice-controls/src/Panel.tsx:5]: I'll remove this.

File: addons/ondevice-controls/src/PropForm.tsx
Comment (dannyhw) [reply addons/ondevice-controls/src/PropForm.tsx:1]: ok I can remove this.

**Pull request merged**
